engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/AbstractTableLocationProvider.java

```diff
@@ -198,18 +198,18 @@ public TableLocation getTableLocationIfPresent(@NotNull final TableLocationKey t
         // See JavaDoc on tableLocations for background.
         // The intent is to create a TableLocation exactly once to replace the TableLocationKey placeholder that was
         // added in handleTableLocationKey.
-        if (current instanceof TableLocation) {
-            return (TableLocation) current;
-        }
-        synchronized (tableLocations) {
-            current = tableLocations.get(tableLocationKey);
-            if (current instanceof TableLocation) {
-                return (TableLocation) current;
+        while (!(current instanceof TableLocation)) {
+            final TableLocationKey immutableKey = (TableLocationKey) current;
+            // noinspection SynchronizationOnLocalVariableOrMethodParameter
+            synchronized (immutableKey) {
+                current = tableLocations.get(immutableKey);
+                if (immutableKey == current) {
+                    // Note, this may contend for the lock on tableLocations
+                    tableLocations.add(current = makeTableLocation(immutableKey));
+                }
             }
-            final TableLocation result = makeTableLocation((TableLocationKey) current);
-            tableLocations.add(result);
-            return result;
         }
+        return (TableLocation) current;
```

Comment:
So, I think there is a slight semantic difference than the previous version. Previously, we were guaranteeing that all calls to `getTableLocationIfPresent` will return the exact same `TableLocation` instance. In this version, it looks like it's technically possible to have two racers both create and return their own results from `makeTableLocation`? If we care to disallow this, maybe we should use `io.deephaven.hash.KeyedObjectHash#putIfAbsent(K, V)` logic (and maybe cleanup loser?)?

Reply:
No. Imagine a scenario with two racers.
The winner sees `current = TLK`, locks TLK, and replaces TLK in the map with TL.
The loser gets one of 3 scenarios:
1. It sees TL on the first read, never locks anything, returns TL.
2. It sees TLK on the first read, saves that to `immutableKey`, locks TLK, refetches, and gets `current = TL`. Since `TL != TLK`, `current != immutableKey`, and we will exit the loop to return the new `current` which is TL.